### PR TITLE
help: document DetectSilence w/ multiple inputs

### PR DESCRIPTION
**Pull request by snappizz**
Please check my work here. This might be totally wrong, or there might be a better way to do this.

### CODE REVIEW

File: HelpSource/Classes/DetectSilence.schelp [hunk 1]
```diff
@@ -77,3 +78,15 @@ Task({
 )
 
 ::
+
+section::DetectSilence and multiple inputs
+
+DetectSilence has a non-obvious interaction with multiple inputs. This code
```
Comment (brianlheim) [inline HelpSource/Classes/DetectSilence.schelp]: I'd avoid saying 'non-obvious' here; this behavior could be perfectly predictably to someone who hasn't used DetectSilence before and is only familiar with multi-channel expansion. "Note that this line" should be enough.

**Comment by brianlheim** [comment]
I think you want `FreeSelf` instead of `FreeSelfWhenDone`. `DetectSilence` doesn't have a 'done' flag detectable by `Done`/`FreeSelfWhenDone` (or if it does, that's not documented, which is another issue). At the very least, when I run your suggested workaround, it doesn't ever free itself.

My question is, couldn't one also use other one-liners like `sig.abs.sum`? That's been my workaround in the past. Obviously it's not a perfect equivalent to the workaround you suggest—in the sense that the threshold value now has a different meaning—but I don't think anyone's being that picky about their uses of `DetectSilence`. Here's a comparison of multiple approaches along with the number of UGens they spawn for an array of size 20.

```supercollider
(
// 1. 49 ugens
// naïve solution, will exit early
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	DetectSilence.ar(sig, doneAction:0);
	sig;
}.play

// 2. 37 ugens
// better, but will exit early if there's near-perfect phase cancellation
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	DetectSilence.ar(sig.sum, doneAction:0);
	sig;
}.play

// 3. 57 ugens
// will not exit even with phase cancellation
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	DetectSilence.ar(sig.abs.sum, doneAction:0);
	sig;
}.play

// 4. 58 ugens
// scaled by the number of channels so that the threshold value makes a little more sense
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	DetectSilence.ar(sig.abs.sum/sig.size, doneAction:0);
	sig;
}.play

// 5. 69 ugens
// most logically sound workaround
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	var done = DetectSilence.ar(sig);
	FreeSelf.kr(done.product);
	sig;
}.play

// 6. 70 ugens
// just to confirm that FreeSelf will actually end the synth
{
	var sig = ([Decay.ar(Impulse.ar(0), 2), Decay.ar(Impulse.ar(0), 1)]!10).flat;
	var done = DetectSilence.ar(sig);
	FreeSelf.kr(done.product);
	sig;
}.play

// 7. 70 ugens
// doesn't free self
{
	var sig = ([Decay.ar(Impulse.ar(0), 2), Decay.ar(Impulse.ar(0), 1)]!10).flat;
	var done = DetectSilence.ar(sig);
	FreeSelfWhenDone.kr(done.product);
	sig;
}.play
)
```

Maybe we could give a few possible solutions and briefly discuss the caveats of each. Either way, thanks--this is a really good tip to document!

**Comment by snappizz** [comment]
Good catch, @brianlheim. I've corrected it and added a note with the full-wave rectification solution.

I don't think we should go into much detail comparing solutions. If the user cares that much about the differences, they're probably knowledgeable enough to make the choice on their own.

**Comment by brianlheim** [comment]
@snappizz Good point, works for me!

File: HelpSource/Classes/DetectSilence.schelp [hunk 2]
```diff
@@ -77,3 +78,17 @@ Task({
...
+
+code::DetectSilence.ar([left, right] + Impulse.ar(0), doneAction: 2)::
+
+spawns strong::two:: DetectSilences, and the enclosing Synth frees itself when left OR right becomes silent. (The code::Impulse.ar(0):: is there to ensure that DetectSilence starts detecting silence right away rather than waiting for non-silence.)
+
+Usually you want to detect silence for both inputs, so what's a good way to change that OR into an AND? The solution is strong::not:: to sum code::(left + right) / 2.sqrt::, because there may be phase cancellation in that sum. Instead, use DetectSilence's output instead of its doneAction, and use FreeSelf to do the freeing. Recall that DetectSilence outputs 1 if silence is detected, and 0 otherwise. Taking the product of multiple DetectSilences will return 1 only if all the DetectSilences are 1. So here is the correct solution to freeing a Synth when both code::left:: and code::right:: first fall silent:
+
+code::FreeSelf.kr(DetectSilence.ar([left, right] + Impulse.ar(0)).product)::
+
+It's also possible to mix the signals with full-wave rectification to avert phase cancellation issues (code::[left, right].abs.sum / 2.sqrt::).
```
Comment (brianlheim) [inline HelpSource/Classes/DetectSilence.schelp]: Just one tiny change - missing newline at the end of file for git's sake.

**Comment by telephon** [comment]
btw. `sig.abs.sum/sig.size` is easier written as `sig.abs.mean`.

**Comment by brianlheim** [comment]
Hah, I completely forgot that while I was testing this.

**Comment by brianlheim** [comment]
@snappizz That makes sense but I'm having trouble verifying for myself. Replacing `sig.abs.sum` with `Mix.ar(sig.abs)` makes no difference in the server load or number of UGens spawned for me, even when I increase the size of the signal array to 200. Am I missing something?

**Comment by telephon** [comment]
Both `sum` and `Mix` perform optimizations behind the scenes. I assumed that the only difference is that `a.sum`, where `a` is not an array, will fail.

(It would be reasonable to implement `aUGen.sum` and `aSimpleNumber.sum` each to return itself).

**Comment by snappizz** [comment]
aha. So Mix is actually unnecessary?

**Comment by telephon** [comment]
Here is a test:

```

x = { |n| SynthDef(\test, { Out.ar(0, Mix({ SinOsc.ar }.dup(n))) }) };
y = { |n| SynthDef(\test, {  Out.ar(0, { SinOsc.ar }.dup(n).sum) }) };

x.(1).children
y.(1).children

x.(2).children
y.(2).children

x.(3).children
y.(3).children

x.(4).children
y.(4).children

x.(5).children
y.(5).children

```

All are the same.

One difference is:

```
x = { |n| SynthDef(\test, { Mix({ SinOsc.ar }.dup(n)) }) };
y = { |n| SynthDef(\test, { { SinOsc.ar }.dup(n).sum }) };

x.(3).children // [ a SinOsc, a SinOsc, a SinOsc, a Sum3 ]
y.(3).children // nil
```
So when using `sum`, the pure ugens (without side effects) are getting optimized away.

For non-pure ugens, the optimization is correct:

```
y = { |n| SynthDef(\test, { { Line.kr }.dup(n).sum }) };
y.(3).children // [ a Line, a Line, a Line ]
```

**Comment by telephon** [comment]
So I would propose that `Mix:new`should be implemented as:

```
*new { |array|
    ^array.asArray.sum
}
```

**Comment by telephon** [comment]
… although the result can sometimes be better when you use `sum`:

```

x = { |n| SynthDef(\test, { Out.kr(0, Mix(f.dup(n))) }) };
y = { |n| SynthDef(\test, {  Out.kr(0, f.dup(n).sum) }) };


f = { |i| [2, 3, 4, 5, 6, Line.kr, SinOsc.ar] @@ i };
x.(6).children.collect { |x| [x, x.rate] }.do(_.postln)
y.(6).children.collect { |x| [x, x.rate] }.do(_.postln)


f = { |i| [ SinOsc.ar, 2, 3, 4, 5, 6, Line.kr, SinOsc.ar] @@ i };
x.(6).children.collect { |x| [x, x.rate] }.do(_.postln)
y.(6).children.collect { |x| [x, x.rate] }.do(_.postln)

``

Here, `sum` just sums numbers instead of making a `Sum3` UGe for them.

**Comment by snappizz** [comment]
Discussion on Mix is cont'd at #2642.

File: HelpSource/Classes/DetectSilence.schelp [hunk 3]
```diff
@@ -77,3 +78,17 @@ Task({
 )
 
 ::
+
+section::DetectSilence and multiple inputs
+
+Be careful feeding multiple inputs into DetectSilence, since multichannel expansion can lead to confusing behavior. This code
+
+code::DetectSilence.ar([left, right] + Impulse.ar(0), doneAction: 2)::
```
Comment (brianlheim) [inline HelpSource/Classes/DetectSilence.schelp]: Did you intend this to appear as runnable code? That only happens if `code::` and the closing `::` are on individual lines:
```
code::
DetectSilence...
::
```

**Comment by brianlheim** [comment]
Ping. @snappizz Do you want to make that change re: `code::`? Are we ready to merge otherwise (I vote yes)?

File: HelpSource/Classes/DetectSilence.schelp (hunk 3 above)
Comment (snappizz) [inline HelpSource/Classes/DetectSilence.schelp]: It is not runnable code since it isn't a complete example.

**Comment by snappizz** [comment]
Ready to merge.

File: HelpSource/Classes/DetectSilence.schelp (hunk 3 above)
Comment (brianlheim) [inline HelpSource/Classes/DetectSilence.schelp]: D'oh

**Pull request merged**